File: src/main/java/org/checkerframework/checker/mungo/annotators/MungoTreeAnnotator.java
package org.checkerframework.checker.mungo.annotators;

import com.sun.source.tree.ClassTree;
import com.sun.source.tree.NewClassTree;
import org.checkerframework.checker.mungo.MungoChecker;
import org.checkerframework.checker.mungo.qual.MungoInfo;
import org.checkerframework.checker.mungo.typecheck.MungoTypeInfo;
import org.checkerframework.framework.type.AnnotatedTypeMirror;
import org.checkerframework.framework.type.treeannotator.TreeAnnotator;

import java.nio.file.Paths;

public class MungoTreeAnnotator extends TreeAnnotator {
  private final MungoChecker checker;

  public MungoTreeAnnotator(MungoChecker checker, MungoAnnotatedTypeFactory atypeFactory) {
    super(atypeFactory);
    this.checker = checker;
  }

  @Override
  public Void visitNewClass(NewClassTree node, AnnotatedTypeMirror annotatedTypeMirror) {
    ClassTree tree = node.getClassBody();
    if (tree != null && !annotatedTypeMirror.hasAnnotation(MungoInfo.class)) {
      // Here we handle anonymous classes because doing this in MungoDefaultQualifierForUseTypeAnnotator is not enough
      // Extract information from class declaration so that the correct annotations can be applied to this instance
      MungoTypeInfo anno = checker.getUtils().visitClassTree(Paths.get(atypeFactory.getVisitorState().getPath().getCompilationUnit().getSourceFile().toUri()), tree);
      if (anno != null) {
        annotatedTypeMirror.replaceAnnotation(anno);
      }
    }
    return null;
  }
}

File: src/main/kotlin/org/checkerframework/checker/mungo/annotators/MungoTreeAnnotator.kt
package org.checkerframework.checker.mungo.annotators

import com.sun.source.tree.*
import org.checkerframework.checker.mungo.MungoChecker
import org.checkerframework.checker.mungo.typecheck.MungoNoProtocolType
import org.checkerframework.checker.mungo.typecheck.MungoNullType
import org.checkerframework.checker.mungo.typecheck.MungoPrimitiveType
import org.checkerframework.framework.type.AnnotatedTypeMirror
import org.checkerframework.framework.type.treeannotator.TreeAnnotator

class MungoTreeAnnotator(private val checker: MungoChecker, private val factory: MungoAnnotatedTypeFactory) : TreeAnnotator(factory) {
  override fun visitLiteral(node: LiteralTree, annotatedTypeMirror: AnnotatedTypeMirror): Void? {
    val ret = super.visitLiteral(node, annotatedTypeMirror)
    val type = when (node.kind) {
      Tree.Kind.NULL_LITERAL -> MungoNullType.SINGLETON
      Tree.Kind.STRING_LITERAL -> MungoNoProtocolType.SINGLETON
      else -> MungoPrimitiveType.SINGLETON
    }
    annotatedTypeMirror.replaceAnnotation(type.buildAnnotation(checker.processingEnvironment))
    return ret
  }

  override fun visitVariable(node: VariableTree, type: AnnotatedTypeMirror): Void? {
    val ret = super.visitVariable(node, type)
    if (type is AnnotatedTypeMirror.AnnotatedDeclaredType) {
      factory.typeApplier.apply(node, type)
    }
    return ret
  }

  override fun visitMemberSelect(node: MemberSelectTree, type: AnnotatedTypeMirror): Void? {
    val ret = super.visitMemberSelect(node, type)
    if (type is AnnotatedTypeMirror.AnnotatedDeclaredType) {
      factory.typeApplier.apply(node, type)
    }
    return ret
  }
}
